public/app/plugins/datasource/cloud-monitoring/components/VisualMetricQueryEditor.tsx

```diff
@@ -13,125 +16,263 @@ import { Alignment } from './Alignment';
-          : PreprocessorType.Rate;
-      const { perSeriesAligner } = getAlignmentPickerData(valueType, metricKind, query.perSeriesAligner, preprocessor);
-
-      // On metric name change reset query to defaults except project name and filters
-      Object.assign(query, {
-        ...defaultTimeSeriesList(datasource),
-        projectName: query.projectName,
-        filters: query.filters,
-      });
-      onChange({
-        ...setMetricType(
-          {
-            ...query,
-            perSeriesAligner,
+  }, [datasource, groupBys, metricType, projectName, refId, crossSeriesReducer, timeRange]);
+
+  useEffect(() => {
+    const loadMetricDescriptors = async () => {
+      if (projectName) {
+        const metricDescriptors = await datasource.getMetricTypes(projectName);
+        const services = getServicesList(metricDescriptors);
+        setMetricDescriptors(metricDescriptors);
+        setServices(services);
+      }
+    };
+    loadMetricDescriptors();
+  }, [datasource, projectName, customStyle, selectStyles.optionDescription]);
+
+  useEffect(() => {
+    const getMetricsList = (metricDescriptors: MetricDescriptor[]) => {
+      const selectedMetricDescriptor = getSelectedMetricDescriptor(metricDescriptors, metricType);
+      if (!selectedMetricDescriptor) {
+        return [];
+      }
+
+      const metricsByService = metricDescriptors
+        .filter((m) => m.service === selectedMetricDescriptor.service)
+        .map((m) => ({
+          service: m.service,
+          value: m.type,
+          label: m.displayName,
+          component: function optionComponent() {
+            return (
+              <div>
+                <div className={customStyle}>{m.type}</div>
+                <div className={selectStyles.optionDescription}>{m.description}</div>
+              </div>
+            );
           },
-          type
-        ),
-        preprocessor,
-      });
-    },
-    [onChange, query, datasource]
-  );
+        }));
+      return metricsByService;
+    };
+
+    const metrics = getMetricsList(metricDescriptors);
+    const service = metrics.length > 0 ? metrics[0].service : '';
+    const metricDescriptor = getSelectedMetricDescriptor(metricDescriptors, metricType);
+    setMetricDescriptor(metricDescriptor);
+    setMetrics(metrics);
+    setService(service);
+  }, [metricDescriptors, getSelectedMetricDescriptor, metricType, customStyle, selectStyles.optionDescription]);
+
+  const onServiceChange = ({ value: service }: SelectableValue<string>) => {
+    const metrics = metricDescriptors
+      .filter((m: MetricDescriptor) => m.service === templateSrv.replace(service))
+      .map((m: MetricDescriptor) => ({
+        service: m.service,
+        value: m.type,
+        label: m.displayName,
+        description: m.description,
+      }));
+    // On service change reset all query values except the project name
+    Object.assign(query, { ...defaultTimeSeriesList(datasource), projectName: query.projectName });
```

Comment:
I don't think you need this assign, in  `if` condition below, `onMetricTypeChange` is doing the same, and in the `else` the query is not being updated anyway :shrug:

Reply:
This is partially true, I can change this to just set the `filters` value to `[]` as the assign in `onMetricTypeChange` conserves the value.